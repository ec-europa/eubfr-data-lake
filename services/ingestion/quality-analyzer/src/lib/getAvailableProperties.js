// @flow

const has = (object, key) =>
  object ? hasOwnProperty.call(object, key) : false;

const specialFields = [
  // monetary
  'total_cost',
  'eu_contrib',
  'private_fund',
  'public_fund',
  'other_contrib',
  // we only need to know if centroid is present
  'centroid',
  // we care only for the coordinates
  'location',
];

/**
 * Takes a object with exceptional structure and returns part of the initial object which is actually useful.
 *
 * @param {Object} initialObject The object representing a special value of a field.
 * @returns {Object} Only a selected part of the {initialObject} which is actually needed.
 *
 * @example
 * input => { "value": 20511539, "currency": "EUR", "raw": "20511539" }
 * output => { "value": 20511539 }
 *
 * @example
 * We care only about the value of a monetary field, new normalized structure is returned.
 *
 * input => { "value": 20511539, "currency": "EUR", "raw": "20511539" }
 * output => { "value": 20511539 }
 *
 * We care only if a centroid has latitude or longitude, `lat` and `lon` fields are skipped.
 *
 * input => { "lat": 52.022163, "lon": 20.914336 }
 * output => { skip: true }
 *
 * We care only if coordinates are present, `coordinates` field is preseved.
 *
 * input => { "type": "Point", "coordinates": [20.914336, 52.022163] }
 * output => { coordinates: true }
 *
 */
const getNormalizedObject = (initialObject: Object): Object => {
  // A monetary field in budget, value is enough of the all
  if (has(initialObject, 'currency')) {
    return { value: initialObject.value };
  }

  // A centroid with latitude and longitude, skip as we care only that it has .centroid
  // This is how we IGNORE the property/field of current level
  if (has(initialObject, 'lat') && has(initialObject, 'lon')) {
    return { skip: true };
  }

  // We care only if there are coordinates or not, not about their values
  // This is how we KEEP the property/field of current level
  if (has(initialObject, 'type') && has(initialObject, 'coordinates')) {
    return { coordinates: true };
  }

  return initialObject;
};

/**
 * Flags whether a given value is useful or not.
 *
 * @param {any} input The input to varify
 *
 * @returns {boolean} Whether acceptable.
 *
 */
const hasUsefulData = input => {
  const unAcceptableValues = [0, '0', 'EUR', '', null];

  if (!unAcceptableValues.includes(input)) {
    return true;
  }

  return false;
};

/**
 * Get information about which fields of an object are not empty.
 * To use, initiate `results` out of the scope of this function and call it on it.
 * Function does not return but accumulates results in `results`.
 *
 * @param {Object} obj Input object to inspect.
 * @param {Array<string>} results The accumulator of results after function's invocation.
 * @param {string} stack The string representation of a nested field in an object as deep as necessary.
 *
 * @returns {boolean} Whether acceptable.
 *
 */
const getAvailableProperties = (
  obj: Object,
  results: Array<string>,
  stack: string = ''
): void => {
  if (!obj) {
    return;
  }

  Object.keys(obj).forEach(property => {
    if (typeof obj[property] === 'object') {
      const fieldIsSpecial = specialFields.includes(property);

      const ObjectUnderInvestigation = fieldIsSpecial
        ? getNormalizedObject(obj[property])
        : obj[property];

      // If no previous stack, then property is sufficient as a root.
      const setStack = stack === '' ? property : `${stack}.${property}`;
      // Re-iterate
      getAvailableProperties(ObjectUnderInvestigation, results, setStack);
    } else if (hasUsefulData(obj[property]) && !has(results, stack)) {
      // nothing nested, make this property the root of the stack for this property
      if (stack === '') {
        results.push(property);
      } else {
        // The stack is the nested fields information
        const setStack = property === 'skip' ? stack : `${stack}.${property}`;
        results.push(setStack);
      }
    }
  });
};

export default getAvailableProperties;
