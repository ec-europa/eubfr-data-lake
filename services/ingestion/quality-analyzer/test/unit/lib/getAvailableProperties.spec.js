/**
 * @jest-environment node
 */

import path from 'path';
import getAvailableProperties from '../../../src/lib/getAvailableProperties';

describe(`Helper getAvailableProperties "@eubfr/ingestion-quality-analyzer" works for any existing transform`, () => {
  const transforms = [
    'agri/csv',
    'budg/xls',
    'iati/csv',
    'inforegio/json',
    'inforegio/xml',
    'valor/xls',
    'wifi4eu/xls',
  ];

  transforms.forEach(transform => {
    test(`Works with results of ${transform}`, () => {
      const results = [];

      // eslint-disable-next-line
      const record = require(path.resolve(
        `${__dirname}/../../stubs/${transform}/record.json`
      ));

      getAvailableProperties(record, results);

      expect(results).toMatchSnapshot();
    });
  });
});
