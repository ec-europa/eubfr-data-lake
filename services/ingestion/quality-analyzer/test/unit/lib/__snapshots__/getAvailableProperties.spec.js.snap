// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Helper getAvailableProperties "@eubfr/ingestion-quality-analyzer" works for any existing transform Works with results of agri/csv 1`] = `
Array [
  "budget.eu_contrib",
  "budget.eu_contrib",
  "budget.funding_area",
  "budget",
  "description",
  "ec_priorities",
  "media.0",
  "media.0",
  "media.0.meta",
  "programme_name",
  "project_id",
  "project_locations.0",
  "project_locations.0.centroid",
  "project_locations.0.centroid",
  "project_locations.0.location",
  "project_locations.0.location.coordinates",
  "project_locations.0.location.coordinates",
  "results",
  "third_parties.0",
  "third_parties.0",
  "timeframe",
  "timeframe",
  "title",
]
`;

exports[`Helper getAvailableProperties "@eubfr/ingestion-quality-analyzer" works for any existing transform Works with results of budg/xls 1`] = `
Array [
  "action",
  "budget.eu_contrib",
  "budget.eu_contrib",
  "call_year",
  "description",
  "programme_name",
  "project_id",
  "project_locations.0",
  "project_locations.1",
  "project_locations.2",
  "project_locations.3",
  "project_locations.4",
  "project_locations.5",
  "project_locations.6",
  "project_locations.7",
  "project_locations.8",
  "project_locations.9",
  "project_locations.10",
  "project_locations.11",
  "project_locations.12",
  "project_locations.13",
  "project_locations.14",
  "results",
  "results",
  "status",
  "sub_programme_name",
  "success_story",
  "third_parties.0",
  "third_parties.0",
  "third_parties.0",
  "third_parties.0",
  "third_parties.0",
  "third_parties.0",
  "third_parties.0",
  "third_parties.1",
  "third_parties.1",
  "third_parties.1",
  "third_parties.1",
  "third_parties.1",
  "third_parties.1",
  "third_parties.1",
  "third_parties.2",
  "third_parties.2",
  "third_parties.2",
  "third_parties.2",
  "third_parties.2",
  "third_parties.2",
  "third_parties.2",
  "third_parties.3",
  "third_parties.3",
  "third_parties.3",
  "third_parties.3",
  "third_parties.3",
  "third_parties.3",
  "third_parties.3",
  "third_parties.4",
  "third_parties.4",
  "third_parties.4",
  "third_parties.4",
  "third_parties.4",
  "third_parties.4",
  "third_parties.4",
  "third_parties.5",
  "third_parties.5",
  "third_parties.5",
  "third_parties.5",
  "third_parties.5",
  "third_parties.5",
  "third_parties.5",
  "third_parties.6",
  "third_parties.6",
  "third_parties.6",
  "third_parties.6",
  "third_parties.6",
  "third_parties.6",
  "third_parties.6",
  "third_parties.7",
  "third_parties.7",
  "third_parties.7",
  "third_parties.7",
  "third_parties.7",
  "third_parties.7",
  "third_parties.7",
  "third_parties.8",
  "third_parties.8",
  "third_parties.8",
  "third_parties.8",
  "third_parties.8",
  "third_parties.8",
  "third_parties.8",
  "third_parties.9",
  "third_parties.9",
  "third_parties.9",
  "third_parties.9",
  "third_parties.9",
  "third_parties.9",
  "third_parties.9",
  "third_parties.10",
  "third_parties.10",
  "third_parties.10",
  "third_parties.10",
  "third_parties.10",
  "third_parties.10",
  "third_parties.10",
  "third_parties.11",
  "third_parties.11",
  "third_parties.11",
  "third_parties.11",
  "third_parties.11",
  "third_parties.11",
  "third_parties.11",
  "third_parties.12",
  "third_parties.12",
  "third_parties.12",
  "third_parties.12",
  "third_parties.12",
  "third_parties.12",
  "third_parties.12",
  "third_parties.13",
  "third_parties.13",
  "third_parties.13",
  "third_parties.13",
  "third_parties.13",
  "third_parties.13",
  "third_parties.13",
  "third_parties.14",
  "third_parties.14",
  "third_parties.14",
  "third_parties.14",
  "third_parties.14",
  "third_parties.14",
  "third_parties.14",
  "third_parties.15",
  "third_parties.15",
  "third_parties.15",
  "third_parties.15",
  "third_parties.15",
  "third_parties.15",
  "third_parties.15",
  "third_parties.16",
  "third_parties.16",
  "third_parties.16",
  "third_parties.16",
  "third_parties.16",
  "third_parties.16",
  "third_parties.16",
  "timeframe",
  "timeframe",
  "title",
]
`;

exports[`Helper getAvailableProperties "@eubfr/ingestion-quality-analyzer" works for any existing transform Works with results of iati/csv 1`] = `
Array [
  "budget.eu_contrib",
  "budget.eu_contrib",
  "description",
  "project_id",
  "project_locations.0",
  "reporting_organisation",
  "timeframe",
  "timeframe",
  "title",
]
`;

exports[`Helper getAvailableProperties "@eubfr/ingestion-quality-analyzer" works for any existing transform Works with results of inforegio/json 1`] = `
Array [
  "budget.total_cost",
  "budget.total_cost",
  "budget.eu_contrib",
  "budget.eu_contrib",
  "budget.funding_area",
  "description",
  "period",
  "project_id",
  "project_locations.0",
  "project_locations.0",
  "project_locations.0.nuts.0",
  "project_locations.0.nuts.0",
  "project_website",
  "themes",
  "themes",
  "themes",
  "third_parties.0",
  "third_parties.0",
  "third_parties.0",
  "third_parties.0",
  "timeframe",
  "timeframe",
  "title",
  "type",
]
`;

exports[`Helper getAvailableProperties "@eubfr/ingestion-quality-analyzer" works for any existing transform Works with results of inforegio/xml 1`] = `
Array [
  "budget.total_cost",
  "budget.total_cost",
  "budget.eu_contrib",
  "budget.eu_contrib",
  "description",
  "period",
  "project_id",
  "project_locations.0",
  "project_locations.0",
  "project_locations.0.nuts.0",
  "project_website",
  "themes",
  "title",
  "type",
]
`;

exports[`Helper getAvailableProperties "@eubfr/ingestion-quality-analyzer" works for any existing transform Works with results of valor/xls 1`] = `
Array [
  "budget.eu_contrib",
  "budget.eu_contrib",
  "call_year",
  "description",
  "programme_name",
  "project_id",
  "project_locations.0",
  "results",
  "results",
  "status",
  "success_story",
  "third_parties.0",
  "third_parties.0",
  "third_parties.0",
  "third_parties.0",
  "third_parties.0",
  "third_parties.0",
  "title",
]
`;

exports[`Helper getAvailableProperties "@eubfr/ingestion-quality-analyzer" works for any existing transform Works with results of wifi4eu/xls 1`] = `
Array [
  "call_year",
  "description",
  "third_parties.0",
  "third_parties.0",
  "project_id",
  "project_locations.0",
  "project_locations.0",
  "project_locations.0",
  "project_locations.0",
  "project_website",
  "timeframe",
  "title",
]
`;
